tiledb/api/examples/reading_incomplete.rs

```diff
@@ -0,0 +1,348 @@
+        .layout(tiledb::query::QueryLayout::RowMajor)?
+        .start_subarray()?
+        .dimension_range_typed::<i32, _>(0, &[1, 4])?
+        .dimension_range_typed::<i32, _>(1, &[1, 4])?
+        .finish_subarray()
+}
+
+fn grow_buffer<T>(b: &mut BufferMut<T>)
+where
+    T: Copy + Default,
+{
+    let capacity = b.as_ref().len();
+    let new_capacity = 2 * capacity;
+    let _ = std::mem::replace(
+        b,
+        BufferMut::Owned(vec![T::default(); new_capacity].into_boxed_slice()),
+    );
+}
+
+/// Handles the incomplete results manually, similar to what might be done with the C API.
+/// Buffers are provided for the query to fill in. Each step fills in as much
+/// as it can, we print the results and then re-submit the query.
+/// The initial buffer sizes are deliberately small to force the NotEnoughSpace
+/// result, which is handled manually by swapping the buffer inside of the RefCell.
+fn read_array_step() -> TileDBResult<()> {
+    println!("read_array_step");
+
+    let init_capacity = 1;
+
+    let tdb = tiledb::context::Context::new()?;
+
+    let rows_output = RefCell::new(OutputLocation {
+        data: BufferMut::Owned(vec![0i32; init_capacity].into_boxed_slice()),
+        cell_offsets: None,
+    });
+    let cols_output = RefCell::new(OutputLocation {
+        data: BufferMut::Owned(vec![0i32; init_capacity].into_boxed_slice()),
+        cell_offsets: None,
+    });
+    let int32_output = RefCell::new(OutputLocation {
+        data: BufferMut::Owned(vec![0i32; init_capacity].into_boxed_slice()),
+        cell_offsets: None,
+    });
+    let char_output = RefCell::new(OutputLocation {
+        data: BufferMut::Owned(vec![0u8; init_capacity].into_boxed_slice()),
+        cell_offsets: Some(BufferMut::Owned(
+            vec![0u64; init_capacity].into_boxed_slice(),
+        )),
+    });
+
+    let mut qq = query_builder_start(&tdb)?
+        .register_raw("rows", &rows_output)?
+        .register_raw("columns", &cols_output)?
+        .register_raw(INT32_ATTRIBUTE_NAME, &int32_output)?
+        .register_raw(CHAR_ATTRIBUTE_NAME, &char_output)?
+        .build();
+
+    loop {
+        let res = qq.step()?;
+        let final_result = res.is_final();
+
+        if let Some((n_a2, b_a2, (n_a1, _, (n_cols, _, (n_rows, _, _))))) =
+            res.unwrap()
+        {
+            let rows =
+                Ref::map(rows_output.borrow(), |o| &o.data.as_ref()[0..n_rows]);
+            let cols =
+                Ref::map(cols_output.borrow(), |o| &o.data.as_ref()[0..n_cols]);
+            let a1 =
+                Ref::map(int32_output.borrow(), |o| &o.data.as_ref()[0..n_a1]);
+
+            let a2 = {
+                let char_output = char_output.borrow();
+                VarDataIterator::new(n_a2, b_a2, &char_output.borrow())
+                    .expect("Expected variable data offsets")
+                    .map(|bytes| String::from_utf8_lossy(bytes).to_string())
+                    .collect::<Vec<String>>()
```

Comment:
Short answer: no

Long answer: collect is necessary with the way this is written, because the `char_output.borrow()` result has to live longer than the iterator.  That won't happen with the `let a2 = { ... }` structure, so to avoid the collect we have to do `let char_output = char_output.borrow(); let a2 = { ... }` instead.